agent/ecs_cni/plugin.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package ecs_cni
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/cihub/seelog"
+	"github.com/containernetworking/cni/libcni"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/pkg/errors"
+)
+
+// NewClient creates a client of ecs_cni which is used to invoke the plugin
+func NewClient(cfg *Config) CNIClient {
+	pluginPath := CNI_PATH
+	if cfg.PluginPath != "" {
+		pluginPath = cfg.PluginPath
+	}
+
+	cniVersion := CNIVersion
+	if cfg.CniVersion != "" {
+		cniVersion = cfg.CniVersion
+	}
+
+	return &cniClient{
+		PluginPath: pluginPath,
+		CNIVersion: cniVersion,
+		Subnet:     ECSSubnet,
+	}
+}
+
+// SetupNS will set up the namespace of container, including create the bridge
+// and the veth pair, move the eni to container namespace, setup the routes
+func (client *cniClient) SetupNS(cfg *Config) error {
+	vethName := defaultEthName
+	if cfg.VethName != "" {
+		vethName = cfg.VethName
+	}
+
+	cns := &libcni.RuntimeConf{
+		ContainerID: cfg.ContainerID,
+		NetNS:       fmt.Sprintf(netnsFormat, cfg.ContainerPID),
+		IfName:      vethName,
+	}
+
+	netConfigList, err := client.constructNetworkConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "setupNS ecs_cni: Construct network configuration failed")
```

Comment:
(multiple occurrences) Error prefix probably should be "task networking" or "cni".